src/vs/workbench/parts/extensions/browser/extensionsActions.ts

```diff
@@ -1133,6 +1134,120 @@ export class ShowWorkspaceRecommendedExtensionsAction extends Action {
 	}
 }
 
+export class InstallWorkspaceRecommendedExtensionsAction extends Action {
+
+	static ID = 'workbench.extensions.action.installWorkspaceRecommendedExtensions';
+	static LABEL = localize('installWorkspaceRecommendedExtensions', "Install All Workspace Recommended Extensions");
+
+	private disposables: IDisposable[] = [];
+	private recommendations: string[] = [];
+	private toInstall: string[] = [];
+	private updatePromise: TPromise<any> = TPromise.as(null);
+
+	constructor(
+		id: string = InstallWorkspaceRecommendedExtensionsAction.ID,
+		label: string = InstallWorkspaceRecommendedExtensionsAction.LABEL,
+		@IWorkspaceContextService private contextService: IWorkspaceContextService,
+		@IViewletService private viewletService: IViewletService,
+		@IExtensionsWorkbenchService private extensionsWorkbenchService: IExtensionsWorkbenchService,
+		@IExtensionTipsService private extensionTipsService: IExtensionTipsService,
+		@IMessageService private messageService: IMessageService
+	) {
+		super(id, label, 'extension-action');
+		this.extensionsWorkbenchService.onChange(() => this.updateToInstall(), this, this.disposables);
+		this.contextService.onDidChangeWorkbenchState(() => this.updateRecommendations(), this, this.disposables);
+		this.updatePromise = this.updateRecommendations();
+	}
+
+	private updateRecommendations(): TPromise<any> {
+		this.enabled = this.contextService.getWorkbenchState() !== WorkbenchState.EMPTY;
+		if (this.enabled) {
+			return this.extensionTipsService.getWorkspaceRecommendations().then(names => {
+				this.recommendations = names;
+				this.updateToInstall();
+			});
+		}
+		return TPromise.as(null);
+	}
+
+	private updateToInstall(): void {
+		const installed = this.extensionsWorkbenchService.local.map(x => x.id.toLowerCase());
+		this.toInstall = this.recommendations.filter(x => installed.indexOf(x.toLowerCase()) === -1);
+		this.enabled = this.toInstall.length > 0;
+	}
+
+	run(): TPromise<any> {
+		this.viewletService.openViewlet(VIEWLET_ID, true)
+			.then(viewlet => viewlet as IExtensionsViewlet)
```

Comment:
Not sure what is the intention here. To show recommended extensions before install or after install. Data will be stale if you do it before and does not makes sense if it is shown after. Since, this is an install action, how about showing Installed extensions after installing?

Reply:
Intention is to show what the recommendations were.
Data being stale is fine, they will show the install status (Installing and Reload)

Reply:
Ah I see.. If I understand correctly.. the flow is as follows:

- Open extensions viewlet and show workspace recommendations
- Install all recommendations there

Since two promises are not dependent, it can happen by the time viewlet show extensions, they are already installed. How about install only after they are shown?

Reply:
On one hand, doing that would be a good user experience.
User sees the viewlet and then the "Install" there changes to "Installing"/"Reload"

On the other hand, this adds a dependency that isnt required. The install will have to wait until all the gallery queries are done

Reply:
Done. https://github.com/Microsoft/vscode/commit/2d0b55460d380b8693b9b9151021d1671f15268f